build.gradle.kts

```diff
@@ -36,18 +36,41 @@ semver {
     // Do not create an empty release commit when running the "releaseVersion" task.
     createReleaseCommit = false
 
+    // Use "RC" instead of "SNAPSHOT" as the default pre-release identifier.
+    defaultPreRelease = "RC"
+
     // Do not let untracked files bump the version or add a "-SNAPSHOT" suffix.
     noDirtyCheck = true
 }
 
 // Only override a default version (which usually is "unspecified"), but not a custom version.
 if (version == Project.DEFAULT_VERSION) {
-    version = semver.semVersion.takeIf { it.isPreRelease }
-        // To get rid of a build part's "+" prefix because Docker tags do not support it, use only the original "build"
-        // part as the "pre-release" part.
-        ?.toString()?.replace("${semver.defaultPreRelease}+", "")
-        // Fall back to a plain version without pre-release or build parts.
-        ?: semver.version
+    val semVersion = semver.semVersion
+
+    version = buildString {
+        // Use the major, minor, and patch version as the version string, for example, "1.2.3".
+        append(semVersion.major)
+        append(".")
+        append(semVersion.minor)
+        append(".")
+        append(semVersion.patch)
+
+        // If the current commit is a pre-release, append the pre-release identifier, for example, "1.2.3-RC1".
+        // Any commit not tagged as a release is considered a pre-release.
+        if (semVersion.isPreRelease) {
+            append("-") // Do not use a "+" because it is not allowed in Docker tags.
+            append(semVersion.preRelease)
+
+            // If the current commit is ahead of the last tag, append the commit count and the SHA, for example,
+            // "1.2.3-RC1.001.sha.0123456".
+            if (semVersion.commitCount > 0) {
+                append(".")
+                append("%03d".format(semVersion.commitCount))
+                append(".sha.")
+                append(semVersion.sha.take(7))
+            }
+        }
```

Comment:
In general, I wonder whether you could achieve the same with just

    semVersion.toInfoVersionString(shaLength = 7).replace('+', '.')

However, from a SemVer perspective, I believe it's [correct to have a '+' before the `<build>` part](https://semver.org/#backusnaur-form-grammar-for-valid-semver-versions).

Reply:
This would work with a small adaption:
```kotlin
val shaLength = if (semVersion.commitCount > 0) 7 else 0
version = semVersion.toInfoVersionString(shaLength = shaLength).replace('+', '-')
```

I found it difficult to understand how the existing functions behave in all possible scenarios, that's why I made the verbose implementation.

>I believe it's [correct to have a '+' before the <build> part](https://semver.org/#backusnaur-form-grammar-for-valid-semver-versions).

I know, but we replaced it with '-' also before because '+' cannot be used in Docker tags. I initially implemented it so that the '+' is only removed for the Docker tag but kept for the app version itself, but you changed that in 6fdc420.

Reply:
> that's why I made the verbose implementation.

Ok, so I guess you prefer to keep it as-is.

There's still the [optional commit message nit](https://github.com/eclipse-apoapsis/ort-server/pull/1185#discussion_r1792882282), though.

Reply:
I first wanted to clarify if we should move the '+' replacement back to the Docker build job before applying any changes.
As I did the manual testing for the simplified code above now as well I would be fine with using it, hoping that the library is stable there.

Reply:
> I first wanted to clarify if we should move the '+' replacement back to the Docker build job before applying any changes.

On the other hand I'm not even sure if we should continue publishing snapshots of the Docker images once we have releases.

Reply:
I have now switched to the simpler implementation but kept the "+" replacement in Gradle.